Add support for PATCH and PUT for edit actions

mkoura opened the pull request:
As described in [documentation](http://manageiq.org/docs/reference/latest/api/overview/http#updating-resources), the PATCH and PUT HTTP methods are supported for updating a resource.

mkoura committed:
Add support for PATCH and PUT for edit actions

cben commented:
_result_processor part looks great, will review actions part soon...

cben commented on src/manageiq_client/api.py:
```diff
@@ -518,10 +516,17 @@ def reload(self):
 
+            # edit actions on entities can be performed using PATCH and PUT as well
+            if action["name"] == "edit" and isinstance(self._obj, Entity):
+                for edit_method in ("patch", "put"):
```
this is fine but I'm just curious: why is this needed for `.edit` and only for it?
IIUC it happens automatically for e.g. `.delete.POST` & `.delete.DELETE`, because API gives us multiple "delete" actions?
Is it an API bug that multiple "edit" methods are not declared by API?

mkoura replied:
Exactly as you said, "delete" with both POST and DELETE are listed in "actions" section of a resource. No actions with PUT and PATCH are listed there, but should be available for editing a resource wherever editing is supported.
It looks like a bug, I'm gonna make a BZ.

cben commented on src/manageiq_client/api.py:
```diff
@@ -577,28 +582,31 @@ def __call__(self, *args, **kwargs):
+        elif self._method == "patch":
+            result = self.api.patch(self._href, *args)
+        elif self._method in ("delete", "put"):
+            result = getattr(self.api, self._method)(self._href, **kwargs)
```
DELETE used to use `query_dict` which I think was effectively `{"resource": kwargs}`, now uses kwargs directly.
does this still work?

actually, does DELETE need a args at all?  Isn't it just url `DELETE /api/:collection/:id` ?

also, is DELETE supported on collection, passing multiple `resources` ids?  It looks like old code supported that but new one doesn't.

mkoura replied:
Nice catch again 👍 ATM it seems body of DELETE is ignored completely, you can send whatever you want and the resource is deleted anyway. I'm passing kwargs just because DELETE can have body according to the HTTP specification, but for our API it's not needed ATM.
DELETE is supported only for single resource, not on collection.

mkoura commented:
@cben Thanks for approval.
I'll reference the BZ in a comment and make a small change for slightly better behavior in case the bug is fixed, please wait with merging.

mkoura committed:
enhancement of Actions objects creation

mkoura commented:
@cben I changed the code so that new Actions are created and added as action methods only when they don't exist yet. This will also prevent unnecessary creation of PATCH and PUT Actions for every "edit" action once BZ 1491336 is fixed.

cben reviewed (approved):
@simon3z this is ready.

I have an idea for some cosmetic reordering but I'll send a separate PR for that.

simon3z commented:
@mkoura @cben I see two problems with the code around this area.

1. No http code is checked (we can postpone this as it's not there now but we really need a plan for that)

2. I don't like the `empty_allowed` flag that is passed around

I think that there are several ways to simplify `_result_processor` so that we don't need the `empty_allowed` flag. One of the possible solutions could involve leveraging point 1 I mentioned above.

Let me know if need more information on how to achieve that but I'd rather see you trying to improve this on your own first.

mkoura committed:
Don't pass empty_allowed

mkoura committed:
check HTTP response status

mkoura commented:
@simon3z @cben I changed the code, please review/comment.

mkoura committed:
new tests and tests edits

mkoura committed:
Add reason to HTTP status error message

cben commented:
ping @simon3z
